### ISSUE TITLE
Lua script per route configuration using XDS Java control plane API

### ISSUE THREAD

**Issue by lorenzogio97** [envoyproxy/envoy#27270]
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: Lua script per route configuration using XDS Java control plane API

*Description*:
I'm using Java XDS API to implement a custom control plane. I read on the documentation how to specify Lua script filter at listener/VH/route level using YAML configuration file. Unfortunatly I didn't find any information on how to recreate this configuration using Java Api. I was able to build some other configuration, but I don't understand how to do this. At the moment, my code is the following:
![image](https://user-images.githubusercontent.com/45814828/237054233-296b81e5-c0e7-46c9-a3fa-5744f7764c73.png)
(Route object is added later in a RouteConfiguration).
How can I specify Lua inline code in this situation? I need this for a specific route.
Thanks in advance for your help.

**Comment by RyanTheOptimist** [comment]
cc:  @mattklein123 @wbpcode

**Comment by lorenzogio97** [comment]
I think I understood how to write the configuration, the configuration dump seems good, but the code is not executed. 
I added the lua filter to the listener because I read on the documentation that lua filter should be set on the listener in order to be overwritten by the one specified inside the route. This is the code for listener.
`ConfigSource.Builder configSourceBuilder = ConfigSource.newBuilder().setResourceApiVersion(ApiVersion.V3);

        ConfigSource rdsSource = configSourceBuilder
                .setApiConfigSource(
                        ApiConfigSource.newBuilder()
                                .setTransportApiVersion(ApiVersion.V3)
                                .setApiType(ApiConfigSource.ApiType.DELTA_GRPC)
                                .addGrpcServices(
                                        GrpcService.newBuilder()
                                                .setEnvoyGrpc(
                                                        GrpcService.EnvoyGrpc.newBuilder()
                                                                .setClusterName("xds_cluster"))))
                .build();

        Lua lua= Lua.newBuilder()
                .setDefaultSourceCode(DataSource.newBuilder()
                        .setInlineString(
                                "function envoy_on_response(response_handle) " +
                                        "  response_handle:headers():add(\"response_body_size\", response_handle:body():length()) " +
                                        "end"
                        )
                        .build())
                .build();

        HttpConnectionManager manager =
                HttpConnectionManager.newBuilder()
                        .setCodecType(HttpConnectionManager.CodecType.AUTO)
                        .setStatPrefix("http-"+bindIPAddress + "-" + bindPort)
                        .setRds(
                                io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.Rds
                                        .newBuilder()
                                        .setConfigSource(rdsSource)
                                        .setRouteConfigName(routeConfigName))
                        .addHttpFilters(
                                io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter
                                        .newBuilder()
                                        .setName("envoy.filters.http.lua")
                                        .setTypedConfig(Any.pack(lua))
                        )
                        .addHttpFilters(
                                io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter
                                        .newBuilder()
                                        .setName("envoy.filters.http.router")
                                        .setTypedConfig(Any.pack(Router.newBuilder().build()))
                        )
                        .build();


        Listener newListener = Listener.newBuilder()
                .setName("http-"+bindIPAddress + "-" + bindPort)
                .setAddress(
                        Address.newBuilder()
                                .setSocketAddress(
                                        SocketAddress.newBuilder()
                                                .setAddress(bindIPAddress)
                                                .setPortValue(bindPort)
                                                .setProtocol(SocketAddress.Protocol.TCP)))
                .addFilterChains(
                        FilterChain.newBuilder()
                                .addFilters(
                                        Filter.newBuilder()
                                                .setName("envoy.http_connection_manager")
                                                .setTypedConfig(Any.pack(manager))
                                )
                )
                .build();`

This is the code of the route:

```
LuaPerRoute lua= LuaPerRoute.newBuilder()
                .setSourceCode(DataSource.newBuilder()
                        .setInlineString(
                                "function envoy_on_request(request_handle)" +
                                "  local headers, body = request_handle:httpCall(" +
                                "  \"orchestrator_cluster\"," +
                                "  {" +
                                "    [\":method\"] = \"POST\"," +
                                "    [\":path\"] = \""+"/migration_feedback/"+username+"/"+proxyId+"\"" +
                                "  }," +
                                "  \"Hello\", "+
                                "  0,"+
                                "  true)" +
                                "end"
                        )
                        .build())
                .build();

        Route.Builder route = Route.newBuilder()
                .setName(username+"-"+prefix)
                .setMatch(
                        RouteMatch.newBuilder()
                                .setPathSeparatedPrefix(prefix)
                                .addHeaders(HeaderMatcher.newBuilder().setName("cookie").setStringMatch(StringMatcher.newBuilder().setContains("authID="+userCookie))))
                .setRoute(
                        RouteAction.newBuilder()
                                .setPrefixRewrite("/")
                                .setCluster(username+"-"+destinationCluster))
                .putTypedPerFilterConfig("envoy.filters.http.lua", Any.pack(lua));

        RouteConfiguration routeConfiguration =  RouteConfiguration.newBuilder()
                .setName(routeConfigName)
                .addVirtualHosts(
                        VirtualHost.newBuilder()
                                .setName(domain)
                                .addDomains("*"+domain)
                                .addRoutes(route)
                )
                .build();
```

I tried both on_request and on_response functions, but in any case it is not executed. Strange because envoy loads the configuration correctly and from config dump it is exactly the same structure of some example reported in documentation.
Some suggestions?

**Comment by wbpcode** [comment]
Could you provide you config dump result? thanks.

**Comment by lorenzogio97** [comment]
This is the Listener generated configuration:
```
{
   "@type": "type.googleapis.com/envoy.admin.v3.ListenersConfigDump",
   "version_info": "2",
   "dynamic_listeners": [
    {
     "name": "http-0.0.0.0-80",
     "active_state": {
      "version_info": "339572199cb2ae53521e14adc9511db4ada14afa2282d64ceae0b833bed88d60",
      "listener": {
       "@type": "type.googleapis.com/envoy.config.listener.v3.Listener",
       "name": "http-0.0.0.0-80",
       "address": {
        "socket_address": {
         "address": "0.0.0.0",
         "port_value": 80
        }
       },
       "filter_chains": [
        {
         "filters": [
          {
           "name": "envoy.http_connection_manager",
           "typed_config": {
            "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager",
            "stat_prefix": "http-0.0.0.0-80",
            "rds": {
             "config_source": {
              "api_config_source": {
               "api_type": "DELTA_GRPC",
               "grpc_services": [
                {
                 "envoy_grpc": {
                  "cluster_name": "xds_cluster"
                 }
                }
               ],
               "transport_api_version": "V3"
              },
              "resource_api_version": "V3"
             },
             "route_config_name": "default"
            },
            "http_filters": [
             {
              "name": "envoy.filters.http.lua",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.extensions.filters.http.lua.v3.Lua",
               "default_source_code": {
                "inline_string": "function envoy_on_response(response_handle)   response_handle:headers():add(\"response_body_size\", response_handle:body():length()) end"
               }
              }
             },
             {
              "name": "envoy.filters.http.router",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.extensions.filters.http.router.v3.Router"
              }
             }
            ]
           }
          }
         ]
        }
       ]
      },
      "last_updated": "2023-05-10T08:02:03.259Z"
     }
    }
   ]
  }
```

The config of the RouteConfiguration:

```
  {
   "@type": "type.googleapis.com/envoy.admin.v3.RoutesConfigDump",
   "dynamic_route_configs": [
    {
     "version_info": "fc7257cd656cd98b4902069ccb9643d0b9c025acfc8ec1a9248e9c94fb4e6907",
     "route_config": {
      "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
      "name": "default",
      "virtual_hosts": [
       {
        "name": "lorenzogiorgi.com",
        "domains": [
         "*lorenzogiorgi.com"
        ],
        "routes": [
         {
          "match": {
           "headers": [
            {
             "name": "cookie",
             "string_match": {
              "contains": "authID=D008E02614701F187DE83EA8B94B2231B88D695E1B4D91CBDE91D44012DC246C"
             }
            }
           ],
           "path_separated_prefix": "/echo/echo"
          },
          "route": {
           "cluster": "lorenzo-echo",
           "prefix_rewrite": "/"
          },
          "typed_per_filter_config": {
           "envoy.filters.http.lua": {
            "@type": "type.googleapis.com/envoy.extensions.filters.http.lua.v3.LuaPerRoute",
            "source_code": {
             "inline_string": "function envoy_on_request(request_handle)  local headers, body = request_handle:httpCall(  \"orchestrator_cluster\",  {    [\":method\"] = \"POST\",    [\":path\"] = \"/migration_feedback/lorenzo/edge3\",    [\":authority\"] = \"orchestrator_cluster\"  },  \"Hello\",   0,  true)end"
            }
           }
          },
          "name": "lorenzo-/echo/echo"
         },
         {
          "match": {
           "headers": [
            {
             "name": "cookie",
             "string_match": {
              "contains": "authID=D008E02614701F187DE83EA8B94B2231B88D695E1B4D91CBDE91D44012DC246C"
             }
            }
           ],
           "path_separated_prefix": "/echo/echo2"
          },
          "route": {
           "cluster": "lorenzo-echo2",
           "prefix_rewrite": "/"
          },
          "typed_per_filter_config": {
           "envoy.filters.http.lua": {
            "@type": "type.googleapis.com/envoy.extensions.filters.http.lua.v3.LuaPerRoute",
            "source_code": {
             "inline_string": "function envoy_on_request(request_handle)  local headers, body = request_handle:httpCall(  \"orchestrator_cluster\",  {    [\":method\"] = \"POST\",    [\":path\"] = \"/migration_feedback/lorenzo/edge3\",    [\":authority\"] = \"orchestrator_cluster\"  },  \"Hello\",   0,  true)end"
            }
           }
          },
          "name": "lorenzo-/echo/echo2"
         }
        ]
       }
      ]
     },
     "last_updated": "2023-05-10T08:27:29.042Z"
    }
   ]
  }
```
The only strange things that I notice is that escape characters remains in the configuration. I don't know if this can be a problem for script being executed. I cannot remove escaping in Java, otherwise i will stop the string. Suggestions?

**Comment by wbpcode** [comment]
![image](https://github.com/envoyproxy/envoy/assets/12389633/78418782-fa88-4376-bc9b-b1fc8be217ef)

**Comment by lorenzogio97** [comment]
The script inside the listener configuration its the same reported here:
https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#:~:text=function%20envoy_on_response(response_handle)%0A%20%20%2D%2D%20Wait%20for%20the%20entire%20response%20body%20and%20add%20a%20response%20header%20with%20the%20body%20size.%0A%20%20response_handle%3Aheaders()%3Aadd(%22response_body_size%22%2C%20response_handle%3Abody()%3Alength())

Even the route script is an adaptation of a code present in the documentation. Maybe the problem are related to escape characters? Did you configure your test using XDS or static configuration?

**Comment by wbpcode** [comment]
> The script inside the listener configuration its the same reported here: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#:~:text=function%20envoy_on_response(response_handle)%0A%20%20%2D%2D%20Wait%20for%20the%20entire%20response%20body%20and%20add%20a%20response%20header%20with%20the%20body%20size.%0A%20%20response_handle%3Aheaders()%3Aadd(%22response_body_size%22%2C%20response_handle%3Abody()%3Alength())
> 
> Even the route script is an adaptation of a code present in the documentation. Maybe the problem are related to escape characters? Did you configure your test using XDS or static configuration?

A static configuration that copy the script string from you example directly.

**Comment by wbpcode** [comment]
After a deep going into the source code, seems it is a bug.

You can get body length first. And call the `add` method of headers after that. Then you will get expected result.

**Comment by lorenzogio97** [comment]
> > The script inside the listener configuration its the same reported here: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#:~:text=function%20envoy_on_response(response_handle)%0A%20%20%2D%2D%20Wait%20for%20the%20entire%20response%20body%20and%20add%20a%20response%20header%20with%20the%20body%20size.%0A%20%20response_handle%3Aheaders()%3Aadd(%22response_body_size%22%2C%20response_handle%3Abody()%3Alength())
> > Even the route script is an adaptation of a code present in the documentation. Maybe the problem are related to escape characters? Did you configure your test using XDS or static configuration?
> 
> A static configuration that copy the script string from you example directly.

I update the configuration using XDS Api. I'm using Docker Envoy/envoy-dev image to run the configuration, so the latest one. I will try to use older version. 




> After a deep going into the source code, seems it is a bug.
> 
> You can get body length first. And call the `add` method of headers after that. Then you will get expected result.
> 
> A more weird thing is seems it's designed to like this. Seems the example in doc should never work. I will try a old version envoy.

So if a write it in two statements it should work? For example the httpCall script inside the route configuration it is a single operation, but it doesn't not work. It seems it is never executed in any case. I tried to add --log-level debug or -l debug into my docker file to see more in detail, but envoy continue to log only INFO and above. But strangely I don't see any error message related to Lua, like the one you posted here.

**Comment by wbpcode** [comment]
After checking the code and old version envoy, I think the doc is alway wrong. You should write code like this:

```
                        function envoy_on_response(response_handle)
                          length = response_handle:body():length()
                          response_handle:headers():add("response_body_size", length)
                        end
```

**Comment by wbpcode** [comment]
Sorry for the annoying trouble. I will update the doc and try to find why we designed it like this type.

**Comment by lorenzogio97** [comment]
> After checking the code and old version envoy, I think the doc is alway wrong. You should write code like this:
> 
> ```
>                         function envoy_on_response(response_handle)
>                           length = response_handle:body():length()
>                           response_handle:headers():add("response_body_size", length)
>                         end
> ```

I'm actually interested on the code with httpCall for my project. This code was a test to be sure that I was not wrong. Do you think this bug has some impact on httpCall I posted on the Route configuration?

**Comment by wbpcode** [comment]
If you didn't hybrid sync and async code in same sentence, it should be ok.
And this problem is must present. So, it would be pretty easy to test your production code to ensure it's safe.

**Comment by lorenzogio97** [comment]
> After checking the code and old version envoy, I think the doc is alway wrong. You should write code like this:
> 
> ```
>                         function envoy_on_response(response_handle)
>                           length = response_handle:body():length()
>                           response_handle:headers():add("response_body_size", length)
>                         end
> ```

I confirm that this code works in listener. Putting length method inside headers():add() doesn't work. This is tested updating the configuration using XDS.

**Comment by lorenzogio97** [comment]
> If you didn't hybrid sync and async code in same sentence, it should be ok. And this problem is must present. So, it would be pretty easy to test your production code to ensure it's safe.

I want to do a simple async httpCall, I'm making more tests.

**Comment by lorenzogio97** [comment]
I made more tests. No way to make  HTTPCall work. This script is specified per route. Since a script is specified per route, the default one on the listener is correctly not executed. I investigated using Wireshark and no packet is generated from httpCall. This is the config dump of the last version (I removed escape char using single quote instead of double into Java strings. From what i read lua does not make difference between ' and ").
```
  {
   "@type": "type.googleapis.com/envoy.admin.v3.RoutesConfigDump",
   "dynamic_route_configs": [
    {
     "version_info": "7173b16a0ac54be14fa4cd95d0db68074f4c1f2b70e82e4973d06939085d7b51",
     "route_config": {
      "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
      "name": "default",
      "virtual_hosts": [
       {
        "name": "lorenzogiorgi.com",
        "domains": [
         "*lorenzogiorgi.com"
        ],
        "routes": [
         {
          "match": {
           "headers": [
            {
             "name": "cookie",
             "string_match": {
              "contains": "authID=6EE7B8828D2A306CB564AC9CDF47CC08E5F71A8B7205F8F21814A57E3247F949"
             }
            }
           ],
           "path_separated_prefix": "/echo/echo"
          },
          "route": {
           "cluster": "lorenzo-echo",
           "prefix_rewrite": "/"
          },
          "typed_per_filter_config": {
           "envoy.filters.http.lua": {
            "@type": "type.googleapis.com/envoy.extensions.filters.http.lua.v3.LuaPerRoute",
            "source_code": {
             "inline_string": "function envoy_on_request(request_handle)  request_handle:httpCall('orchestrator_cluster',{[':method'] = 'POST',[':path'] = '/migration_feedback/lorenzo/edge2',},'Hello', 0,true) end"
            }
           }
          },
          "name": "lorenzo-/echo/echo"
         },
         {
          "match": {
           "headers": [
            {
             "name": "cookie",
             "string_match": {
              "contains": "authID=6EE7B8828D2A306CB564AC9CDF47CC08E5F71A8B7205F8F21814A57E3247F949"
             }
            }
           ],
           "path_separated_prefix": "/echo/echo2"
          },
          "route": {
           "cluster": "lorenzo-echo2",
           "prefix_rewrite": "/"
          },
          "typed_per_filter_config": {
           "envoy.filters.http.lua": {
            "@type": "type.googleapis.com/envoy.extensions.filters.http.lua.v3.LuaPerRoute",
            "source_code": {
             "inline_string": "function envoy_on_request(request_handle)  request_handle:httpCall('orchestrator_cluster',{[':method'] = 'POST',[':path'] = '/migration_feedback/lorenzo/edge2',},'Hello', 0,true) end"
            }
           }
          },
          "name": "lorenzo-/echo/echo2"
         }
        ]
       }
      ]
     },
     "last_updated": "2023-05-11T09:32:43.220Z"
    }
   ]
  }
```

**Comment by wbpcode** [comment]
Please add `:authority` header in request.

**Comment by lorenzogio97** [comment]
I have upstream cluster orchestrator_cluster is defined using IP address. Should I add it also in that case?

**Comment by lorenzogio97** [comment]
Ok, I have added it and it works. The request does not appear in logs because I made an error on upstream configuration. I added http2_protocol_options in cluster configuration, but the server supports only http/1.1, so it refuses the request. 
Escape characters  seems to not be an issue, so no problem in writing a script inside a Java string. 
Regarding the original topic, the code that I posted [here](https://github.com/envoyproxy/envoy/issues/27270#issuecomment-1540967910) is the right way to configure Lua scripts in Listener and per Route using Xds Java API.
The only real problem/bug is the one that i quote below. 
Thanks @wbpcode for the help. 

> After checking the code and old version envoy, I think the doc is alway wrong. You should write code like this:
> 
> ```
>                         function envoy_on_response(response_handle)
>                           length = response_handle:body():length()
>                           response_handle:headers():add("response_body_size", length)
>                         end
> ```
